### ISSUE TITLE
Let user select elements by clicking them on the canvas

### ISSUE THREAD

**Issue by alexjlockwood** [romannurik/AndroidIconAnimator#83]
Have you thought at all about this?

For stroked paths, I was thinking given some mouse click at (x,y) we could approximate the shortest distance from that point to each path, and select the path if the distance is less than or equal to half the stroke width.

I'm not so sure how to do it for filled paths though...

In order to account for transformations that may have been applied on the paths, I assume we just need to prebake the transformations into the paths beforehand or something so that we can represent each path as they appear to the user.

I'm new to the web, so I really have no clue what libraries exist out there to do this type of math (or if it is supported natively somehow). Any thoughts?

**Comment by romannurik** [comment]
Definitely thought about this; ideally this would be part of a larger canvas area overhaul, related to #19

**Comment by alexjlockwood** [comment]
I think I agree with what you said before about representing things under-the-hood as approximated bezier curves.

The main reason is because this [bezier-js](https://pomax.github.io/bezierjs/) library exists and seems like it would have pretty much everything we need. To calculate the closest distance from a mouse click to a stroked path, we can use `new Bezier(strokedPathBeziers).project(point)`. To determine whether a mouse click is inside of a closed/filled path... well, that might still be tricky. According to [this primer on bezier curves](https://pomax.github.io/bezierinfo/#shapes) the traditional algorithm that is used is to draw a line from the mouse pointer to an arbitrary point that we know is not contained inside of the path. If that line intersects with the closed path an odd number of times, it is contained inside the path. If the number of intersections is even, it is contained outside the path. We'd use `new Bezier(filledPathBeziers).intersects(line)` in this case.

The other reason is because I think most users aren't going to be able to understand how to edit an elliptical arc command's coordinates anyway... in other words, I feel like it would only ever make sense to allow the user to edit a path by altering bezier curve coordinates (and never by allowing them to alter sweep angles, large arc flags, etc.). What we could probably do is convert the arcs to bezier curves and then allow the user to make changes by dragging the control/end points... it might even be possible to convert the altered bezier curves back to an elliptical arc (there's a bezier-js function to do that)... however, I bet even this is overkill and might not yield great results.

**Comment by alexjlockwood** [comment]
And BTW... do you really think it will be difficult to deal with transformations? Sure it will require rewriting things, but conceptually it seems pretty simple. We just need to make sure the PathLayers have an easy way of knowing their ancestors' transformations, right? Or perhaps we can precalculate the data we'll need at runtime by baking these transformations directly into the PathLayers beforehand. I guess the downside there is that making modifications to groups near the root of the tree might become expensive (since it will require updating all of the paths below it)... but IDK... seems like it is definitely doable somehow.

**Comment by alexjlockwood** [comment]
Here is a video illustrating a hacked up demo I wrote today: https://youtu.be/RlS_MPRIQN8

Similar to how the canvas's `drawLayer_` function that performs canvas operations recursively before drawing the paths, I created a function that recursively traverses the artwork tree, keeping track of the transformation matrices for each inner node and applying them at the leaves to the user's mouse point so that it lives in the transformed coordinate system. In addition to the existing command code inside `svgpathdata.js`, I approximated corresponding `Bezier` curve objects which  the user can click and drag in order to modify the currently displayed path.

The second part of the video shows how we can calculate the shortest distance from a point to a particular path using the `bezier-js` library.

Anyway, it took me most of the day to get this working (mostly because of dumb matrix-related bugs, sighs), but now I have a pretty good idea of what needs to get done. :D 

Also in case you were wondering, yes it works with multiple paths and also works when parent nodes apply transformations on the paths... but it is all pretty messy so gonna rewrite the whole thing tomorrow to see if I can write something more reusable.
